internal/server/server.go

```diff
@@ -120,6 +120,7 @@ func (s *Server) buildHandler() http.Handler {
 	var handler http.Handler
 
 	handler = s.router
+	handler = WithRequestStartMiddleware(handler)
```

Comment:
You probably should add this before or after the `WithRequestIDMiddleware()` handler so that this is injected as early as possible, given that handlers are executed in the inverse order.

An option would be to augment the `WithRequestIDMiddleware()` handler with the new header it would inject.

Reply:
Awesome, didn't realize they get executed in reverse. Wanted to add this one first. Thank you.

Reply:
No worries. You can see this with a simpler example:

```go
  package main

  import (
          "fmt"
          "net/http"
  )

  func One(h http.Handler) http.Handler {
          return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
                  fmt.Println("One")
                  h.ServeHTTP(w, r)
                  fmt.Println("One")
          })
  }

  func Two(h http.Handler) http.Handler {
          return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
                  fmt.Println("Two")
                  h.ServeHTTP(w, r)
                  fmt.Println("Two")
          })
  }

  func Main(h http.Handler) http.Handler {
          return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
                  fmt.Println("Main")
          })
  }

  func main() {
          mux := http.NewServeMux()

          handler := Main(mux)
          handler = Two(handler)
          handler = One(handler)

          mux.Handle("/", handler)
          panic(http.ListenAndServe(":3000", mux))
  }
```

Then, sending a single request against TCP/3000, per:

```console
$ go run main.go
^Z
[1]+  Stopped                 go run main.go
$ bg
[1]+ go run main.go &
$ curl -v 0:3000
*   Trying 0.0.0.0:3000...
* Connected to 0.0.0.0 (0.0.0.0) port 3000
* using HTTP/1.x
> GET / HTTP/1.1
> Host: 0.0.0.0:3000
> User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
> Accept: */*
>
* Request completely sent off
One
Two
Main
Two
One
< HTTP/1.1 200 OK
< Date: Sun, 24 Nov 2024 15:40:48 GMT
< Content-Length: 0
<
* Connection #0 to host 0.0.0.0 left intact
$ fg
go run main.go
^Csignal: interrupt
```

You chain the execution of each handler, which becomes a callback, so to speak, which runs the handler you passed into it, hence the inverse order in which these run. I hope this makes sense.

There are some nice packages such as the [justinas/alice](https://github.com/justinas/alice) that aim to make this pattern a little bit more grounded.